feat: add mapping to immutable sequences

TimothyMakkison opened the pull request:
Partially resolving #250 and #251.

- Added `ImmutableList` and `ImmutableArray`
- Added `EnsureCapacity` generator to `ForEachAddEnumerableMapping`, it works on `ICollection` and `ICollection<T>` and casts the sequence in case the interface was implemented explicitly (aka arrays).
- Added `AddRange` for existing `List`
- Add `IsArray` guard to `TryBuildExistingTargetMapping`, Array explicitly implements a dummy `IList.Add` method that thows `NotImplementedException`. as a result mapperly will attempt to generate a foreach loop adding items to the array.

As `ImmutableList` and `ImmutableArray` cannot be mutated without a setter method, should `EnumerableMappingBuilder.TryBuildExistingTargetMapping` report an error instead of returning null?

I'm not 100% happy with how I added `EnsureCapacity` it will have to be revisited when being used for dictionaries. A future change could use `Enumerable.TryGetNonEnumeratedCount` instead.

latonz commented on src/Riok.Mapperly/Descriptors/Mappings/LinqAddRangeMapping.cs:
```diff
@@ -0,0 +1,50 @@
+        }
+        else
+        {
+            mappedSource = _elementMapping.Build(ctx);
```
This seems to be uncovered by tests, could you add a test which covers this?

latonz commented on src/Riok.Mapperly/Descriptors/MappingBuilders/EnumerableMappingBuilder.cs:
```diff
@@ -58,14 +65,30 @@ public static class EnumerableMappingBuilder
+            return null;
+        }
+
+        if (ctx.Target.ImplementsGeneric(ctx.Types.List, out _))
```
Is a `Select` + `AddRange` mapping faster than a `foreach` + `Add` mapping (assumed https://github.com/riok/mapperly/issues/251 is implemented)? It makes sense to me if the `elementMapping` is synthetic, but does it also make sense for non-synthetic mappings?

latonz reviewed (changes requested):
Thank you for another great contribution!
Could you split the support for the immutable array/list (https://github.com/riok/mapperly/issues/250) and EnsureCapacity (https://github.com/riok/mapperly/issues/251) into two separate PR's? Separating features into single PR's makes reviews a lot easier and the changes are easier to track.

TimothyMakkison commented:
Thanks for the code review, I'll get round to creating a separate EnsureCapacity pull request.

- Added some support for some Immutable type.
- Created `HasImplicitInterfaceMethod` to prevent mapperly calling ICollection<T>.Add on Immutable types when they have an explicit implementation - related to #286. I can remove this and manually check if you don't like this.
- Added test to ensure that mapperly does not attempt to call ICollection<T>.Add on an array type.
- Added integraton tests, it looks like I only generated values for `Roslyn_4_5`. How would I run it for other versions to be like #300? I'm not sure why `ImmutableArrayValue` `ImmutableQueueValue` and `ImmutableStackValue` are the only immutable types that appear inside `RecursiveObject` [here](https://github.com/riok/mapperly/pull/305/files#diff-3d68b0b1676cb425b7402a21499bce6b1f410c02965625099f06f2822f105179).
```C#
// RecursiveObject...
ImmutableArrayValue: null,
ImmutableQueueValue: [],
ImmutableStackValue: [],
```
- Update names of generic types in `WellKnowTypes`

latonz commented on src/Riok.Mapperly/Descriptors/MappingBuilders/EnumerableMappingBuilder.cs:
```diff
@@ -128,30 +148,77 @@ private static (bool CanMapWithLinq, string? CollectMethod) ResolveCollectMethod
+
+    private static IMethodSymbol? ResolveImmutableCollectMethod(MappingBuilderContext ctx)
+    {
+        if (ctx.Target.ImplementsGeneric(ctx.Types.ImmutableArrayT, out _))
```
First I was wondering why this works as `ImplementsGeneric` expects the `genericInterfaceSymbol` parameter to be an interface. It does actually work, since `ImplementsGeneric` does have a `SymbolEqualityComparer.Default.Equals(t.OriginalDefinition, genericInterfaceSymbol)` check at the beginning of the implementation.
I would consider this an implementation detail of the  `ImplementsGeneric` method and instead of calling this method replace these calls with `SymbolEqualityComparer.Default.Equals(ctx.Target.OriginalDefinition, ctx.Types.ImmutableArrayT)`.

latonz commented on src/Riok.Mapperly/Helpers/SymbolExtensions.cs:
```diff
@@ -78,6 +78,32 @@ internal static IEnumerable<IPropertySymbol> GetAllAccessibleProperties(this ITy
     }
 
+
+    // TODO: refactor and rename.
```
Resolve the todo,
nit: two empty lines instead of one before the method...

latonz commented on src/Riok.Mapperly/Helpers/SymbolExtensions.cs:
```diff
@@ -78,6 +78,32 @@ internal static IEnumerable<IPropertySymbol> GetAllAccessibleProperties(this ITy
+
+        // if null then the method is unimplemented
+        if (methodInterImplementaton is null)
+            return true;
```
Should this method really return true in case the method is not implemented at all? `HasImplicitInterfaceMethod` resolving to true for unimplemented methods seems unexpected to me. But this can never happen anyway (due to the `ImplementsGeneric` check on L92, can it? So probably just an exception could be thrown?

TimothyMakkison replied:
Spot on, ordinarily you'd expect this method to be implemented as part of the underlying interface. But in cases like:
```C#
public class RepeatedField : ICollection<int> { }
```
The class `RepeatedField` "implements" the interface but hasn't implemented any methods. This normally fails to compile but except for the unit tests where the above example is common.

My solution is hacky but its better than updating all the tests. I'll add a comment explaining why I've done this.

latonz commented on test/Riok.Mapperly.IntegrationTests/BaseMapperTest.cs:
```diff
@@ -79,6 +81,12 @@ protected TestObject NewTestObj()
                 SourceTargetSameObjectType = new TestObject(8) { IntValue = 99, RequiredValue = 98, },
                 StackValue = new Stack<string>(new[] { "1", "2", "3" }),
                 QueueValue = new Queue<string>(new[] { "1", "2", "3" }),
+                ImmutableArrayValue = ImmutableArray.Create(new[] { "1", "2", "3" }),
```
These array creations are redundant as the target parameter is a `params`...

latonz reviewed (changes requested):
Thank you for the updates 😊 
* I'm fine with the `HasImplicitInterfaceMethod` check 👍 
* Integration tests: Once the github action integration tests for net6.0 and net framework have run, the received snapshot files can be downloaded from the [artifacts](https://github.com/riok/mapperly/actions/runs/4579113048?pr=305). If you wan't to run the tests locally you can either try to use [act](https://github.com/nektos/act) (I personally never tried that) or run the commands of the test github action manually (see [test.yml github action](https://github.com/riok/mapperly/blob/main/.github/workflows/test.yml#L46)).
* I don't know why the other values don't show up in the integration test snapshots... Probably it is some serialization logic of VerifyTests. If you assign a value, it will show up in the snapshot.

TimothyMakkison commented on src/Riok.Mapperly/Descriptors/MappingBuilders/EnumerableMappingBuilder.cs:
```diff
@@ -58,13 +72,16 @@ public static class EnumerableMappingBuilder
+        // create a foreach loop with add calls if source is not an array
+        // and  ICollection.Add(T): void is implemented and not explicit
+        // ensures add is not called and immutable types
+        if (!ctx.Target.IsArrayType() && ctx.Target.HasImplicitInterfaceMethod(ctx.Types.ICollectionT, nameof(ICollection<object>.Add)))
```
`HasImplicitInterfaceMethod` uses `FindImplementationForInterfaceMember` to get the matching interface method. For some reason this returns null if the symbol is an array type. 

This could be due to how interfaces work with arrays generics, I dont think `GetMembers` works on array types.
```C#
// if the symbol is an array.
symbol.GetMembers().Count() == 0;
```

TimothyMakkison commented:
Thanks for the suggestions, I'll try to add the integration files for 4_4 and under.

I've noticed a slight oversight, the [tests for .Net framework 4.8](https://github.com/riok/mapperly/actions/runs/4579113048/jobs/8093723358) always fails because [Immutable types](https://github.com/riok/mapperly/actions/runs/4579113048/jobs/8093723358) are .Net core only. I'm a little suprised I could use them in a .Net Standard 2.0 project.

Not sure how to resolve this, my first instinct would be to multi target .Net Standard 2.0 and .Net Core and use preprocessor directives. Thoughts?

latonz commented on test/Riok.Mapperly.Tests/Mapping/EnumerableTest.cs:
```diff
@@ -633,6 +759,78 @@ public Task MapToExistingQueueShouldWork()
     }
 
+    [Fact]
+    public Task EnumerableToReadonlyImmutableList()
```
Not sure if this and the following tests make sense.... They are trying to map from an `int` property to an `Immutable*`, shouldn't the source property type be of type `IEnumerable<int>`?

latonz commented on test/Riok.Mapperly.Tests/Mapping/EnumerableTest.cs:
```diff
@@ -692,4 +902,16 @@ public void EnumerableMappingDisabledShouldDiagnostic()
+            "class A { public ICollection<int> Value { get; } }",
+            "class B { public int[] Value { get; } }");
+
+        return TestHelper.VerifyGenerator(source);
```
I would adjust the name of the test method to `EnumerableToReadOnlyArrayPropertyShouldDiagnostic` and add a comment (eg. `cannot create a mapping which maps to a read only array`).
Probably also change the type of assertion from snapshot to the fluent assertion style instead of the snapshot (`TestHelper.GenerateMapper(source, TestHelperOptions.AllowInfoDiagnostics) .Should()...`), since the snapshot content isn't the thing being tested here.

latonz reviewed (changes requested):
Thanks for the updates 😊

To resolve the issue of the immutable types not being available in net framework, adding the [System.Collections.Immutable nuget package](https://www.nuget.org/packages/System.Collections.Immutable/) to `Riok.Mapperly.IntegrationTests.csproj` as a conditional dependency (in the same item group as `Portable.System.DateTimeOnly`) should help.

TimothyMakkison commented:
Thanks for the review. I've added and updated the tests, changed `HasImplicitMethod` and added the conditional dependency.

I'm not familiar with XUnit verify does it automatically delete snapshot data when the corresponding method is removed or renamed?

latonz reviewed (changes requested):
VerifyTests does not automatically delete unused snapshot files, they need to be cleaned up manually.

TimothyMakkison commented:
Got the integration tests to work 😅 removed the unused snapshot data, refactored the test and did some code cleanup.
Should be ready to merge 🤞

latonz reviewed (changes requested):
The integration tests, especially the ones for the older Roslyn versions are indeed a hassle. If you have an idea how the situation could be improved, I would be glad to hear it 😊 

Looks good now, could you rebase onto main?
